### PR TITLE
[PT-D][Tensor Parallel] Implement Pairwise Parallel and _parallelize_mlp

### PR DESCRIPTION
**Pull request by fduwjj**
As title

### CODE REVIEW

File: spmd/tensor/parallel/style.py [hunk 1]
```diff
@@ -27,6 +21,29 @@ class ParallelStyle(ABC):
     _prepare_input: Optional[_Prepare_Input_Func_Type]
```
Comment (wanchaol) [inline spmd/tensor/parallel/style.py:21]: should the prepare_input/output be non-optional type? I think every parallel style should have them specified? This also matches with the `__init__` function where every style must pass those two function?

File: spmd/tensor/parallel/utils.py [hunk 2]
```diff
@@ -104,3 +103,42 @@ def wrapper(*args, **kwargs):  # pyre-ignore[2, 3]
...
+def _create_1d_device_mesh(
+    device_mesh: DeviceMesh, tp_mesh_dim: int = 0
+) -> DeviceMesh:
+    """
+    This function converts a N-D ``device_mesh`` into a 1D ``device_mesh``
+    for 1D Tensor Parallelism.
+
+    Args:
+        device_mesh (DeviceMesh): :class:``DeviceMesh`` object which contains
+            how we distribute tensor across GPUs.
```
Comment (wanchaol) [inline spmd/tensor/parallel/utils.py:117]: nit: device_mesh which describes the mesh topology of devices for the DTensor

File: spmd/tensor/parallel/utils.py [hunk 3]
```diff
@@ -104,3 +103,42 @@ def wrapper(*args, **kwargs):  # pyre-ignore[2, 3]
...
+        tp_mesh_dim (int): the dimension of ``device_mesh`` where we perform
+            Tensor Parallelism on.
+
+    Return:
+        device_mesh (DeviceMesh): :class:``DeviceMesh`` object which contains
```
Comment (wanchaol) [inline spmd/tensor/parallel/utils.py:122]: sth like: a 1-D :class:``DeviceMesh`` that Tensor Parallelism works on

File: spmd/tensor/parallel/api.py [hunk 4]
```diff
@@ -86,3 +87,144 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+    ):
+        if i % 2 == 0:
+            # Col-wise Parallelize the linear layer
+            _distribute_linear_module(
+                m,  # pyre-ignore[6]  # type: ignore[arg-type]
+                device_mesh,
+                [Shard(0)],
+                [Shard(0)],
+            )
+            m.register_forward_pre_hook(
```
Comment (wanchaol) [inline spmd/tensor/parallel/api.py:209]: shouldn't we only register pre_hook on the first layer, and register post hook on the last layer? If `_parallel_mlp` works on 8 linear layers, we should only register hook for module 0 and module 7?

Also, I'm wondering if we could directly use `distribute_module` on MLPModule, as that'll only register prehook and posthook once, and maybe it also makes easier for pre/post hook?

File: spmd/tensor/parallel/api.py [hunk 5]
```diff
@@ -86,3 +87,144 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+        )
+
+    if not _has_even_num_linears(module):
+        raise RuntimeError("We only support even number of Linear for MLP.")
+
+    if device_mesh.ndim > 1:
+        device_mesh = _create_1d_device_mesh(device_mesh, tp_mesh_dim)
+
+    for i, m in enumerate(
+        filter(lambda x: isinstance(x, nn.Linear), module.children())
```
Comment (wanchaol) [inline spmd/tensor/parallel/api.py:199]: I'm wondering if we should ensure sth weird not happen: i.e. `linear0 -> emb -> linear1` will be treated as a legit one but it's actually not, maybe we should check the layers in the middle and ensure they don't have parameters?

File: spmd/tensor/parallel/style.py [hunk 6]
```diff
@@ -27,6 +21,29 @@ class ParallelStyle(ABC):
...
+        self._prepare_input = _prepare_input
+        self._prepare_output = _prepare_output
+
+
+class PairwiseParallel(ParallelStyle):
+    """
+    We concatenate colwise and rowwise styles as a fixed pair like
+    what Megatron-LM(https://arxiv.org/abs/1909.08053) is doing.
+    We assume both input and output needs to a replicate DTensor.
+    We now only support this style for Multihead Attention, MLP and
```
Comment (wanchaol) [inline spmd/tensor/parallel/style.py:35]: this line is kinda duplicate with the warning? Maybe we should only keep one.

File: spmd/tensor/parallel/style.py [hunk 7]
```diff
@@ -27,6 +21,29 @@ class ParallelStyle(ABC):
...
+    transformer model.
+
+    .. warning::
+        We need to assume the module to be distributed as a
```
Comment (wanchaol) [inline spmd/tensor/parallel/style.py:39]: Maybe it's not something we "assume", it's something this ParallelStyle requires?

File: spmd/tensor/parallel/style.py [hunk 8]
```diff
@@ -27,6 +21,29 @@ class ParallelStyle(ABC):
...
 
+    @abstractmethod
+    def __init__(self, _prepare_input, _prepare_output) -> None:
+        self._prepare_input = _prepare_input
+        self._prepare_output = _prepare_output
+
+
+class PairwiseParallel(ParallelStyle):
+    """
+    We concatenate colwise and rowwise styles as a fixed pair like
```
Comment (wanchaol) [inline spmd/tensor/parallel/style.py:32]: "PairwiseParallel style concatenates colwise ..."

File: test/spmd/tensor/parallel/test_parallelize_api.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+            .reshape(
+                self.world_size // (self.world_size // tp_size),
+                self.world_size // tp_size,
+            )
+            .to(torch.int)
+        )
+        mesh = DeviceMesh(self.device_type, mesh_shape)
+        one_dimention_mesh_shape = mesh_shape[self.rank // tp_size, :]
+        pg = mesh.get_dim_groups()[1]
+        new_mesh = _create_1d_device_mesh(mesh, 1)
```
Comment (wanchaol) [inline test/spmd/tensor/parallel/test_parallelize_api.py:43]: can we also add a case to test tp_dim=0?

**Review by wanchaol — COMMENTED** [review]
Awesome work! Mostly some suggestions about comment, and have some questions about parallel_mlp works on some weird module pattern and hook registration, please see comments inlined.

File: spmd/tensor/parallel/api.py [hunk 10]
```diff
@@ -86,3 +87,144 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+            we prepare input/output for Tensor Parallelism.
+        tp_mesh_dim (int):
+            the dimension of ``device_mesh`` where we perform
+            Tensor Parallelism on.
+
+    Return:
+        None
+
+    .. warning::
+        We now only support ``PairwiseParallel`` for now.
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply spmd/tensor/parallel/api.py:177]: You are right. But we might have other style for MLP parallelization in the future, right?

File: spmd/tensor/parallel/api.py [hunk 11]
```diff
@@ -86,3 +87,144 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+    """
+
+    # Define hook functions needed for preparing Input/Output.
+    def _module_forward_pre_hook(*args):  # pyre-ignore[2, 3]
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply spmd/tensor/parallel/api.py:181]: yes...

File: spmd/tensor/parallel/style.py (hunk 1 above)
Comment (fduwjj) [reply spmd/tensor/parallel/style.py:21]: Actually that's a good point, let me remove it.

File: spmd/tensor/parallel/api.py (hunk 5 above)
Comment (fduwjj) [reply spmd/tensor/parallel/api.py:199]: But softmax and layerNorm also have parameters. It's hard to differentiate it from embedding in a generic way.

File: spmd/tensor/parallel/style.py (hunk 6 above)
Comment (fduwjj) [reply spmd/tensor/parallel/style.py:35]: Oh sure.

File: test/spmd/tensor/parallel/test_parallelize_api.py (hunk 9 above)
Comment (fduwjj) [reply test/spmd/tensor/parallel/test_parallelize_api.py:43]: Sure.

File: spmd/tensor/parallel/api.py (hunk 4 above)
Comment (fduwjj) [reply spmd/tensor/parallel/api.py:209]: Hmmm I don't have strong opinion for whether to use `distribute_module` or not. But the thing is that how are we going to write `partition_fn`? Unless we also do this layer by layer.

File: spmd/tensor/parallel/api.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (fduwjj) [reply spmd/tensor/parallel/api.py:181]: Use `distribute_module`

File: spmd/tensor/parallel/style.py (hunk 1 above)
Comment (fduwjj) [reply spmd/tensor/parallel/style.py:21]: NVM has a solution here.

File: spmd/tensor/parallel/api.py (hunk 4 above)
Comment (fduwjj) [reply spmd/tensor/parallel/api.py:209]: NVM, found a solution.

File: spmd/tensor/parallel/api.py [hunk 12]
```diff
@@ -86,3 +87,112 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+    Args:
+        module (nn.Module):
+            :class:``nn.Module`` object to be traversed and counted.
+
+    Return:
+        A boolean object which specifies whether the module contains
+        event-number of Linears in its children.
+
+    .. warning::
+        The traversal is not recursive for now.
```
Comment (wanchaol) [inline spmd/tensor/parallel/api.py:106]: qq: why we need this function to be recursive?

File: spmd/tensor/parallel/style.py (hunk 1 above)
Comment (wanchaol) [reply spmd/tensor/parallel/style.py:21]: I see that we didn't remove the `Optional`, so I assume there're still cases when a parallel style have `_prepare_input` or `_prepare_output` be None? Could you share more about those cases?

File: spmd/tensor/parallel/utils.py [hunk 13]
```diff
@@ -44,6 +43,9 @@ def _prepare_input_validate(
     def wrapper(*args, **kwargs):  # pyre-ignore[2, 3]
         assert len(args) >= 1, "_prepare_input needs at least one arg."
         input = args[0]
+        if isinstance(input, list) or isinstance(input, tuple):
+            input = input[0]
+            args = (input, *args[1:])
```
Comment (wanchaol) [inline spmd/tensor/parallel/utils.py:48]: so if the first arg is a list of tuple, we only convert the first item in the list/tuple to DTensor? What if the first arg is sth like `(tensor, tensor, tensor)`?

File: spmd/tensor/parallel/style.py [hunk 14]
```diff
@@ -27,6 +21,27 @@ class ParallelStyle(ABC):
...
+
+
+class PairwiseParallel(ParallelStyle):
+    """
+    PairwiseParallel concatenate colwise and rowwise styles as a fixed
+    pair like what Megatron-LM(https://arxiv.org/abs/1909.08053) is doing.
+    We assume both input and output needs to a replicate DTensor.
+
+    .. warning::
+        We need to the module to be distributed as a
```
Comment (wanchaol) [inline spmd/tensor/parallel/style.py:37]: looks like the some typos, i'm wondering if this would make it more clear: "PairwiseParallel only supports ``nn.Multihead Attention``, ``nn.Transformer`` or MLP Layers (i.e. even number ``nn.Linear`` layers) for now"

File: spmd/tensor/parallel/api.py (hunk 12 above)
Comment (fduwjj) [reply spmd/tensor/parallel/api.py:106]: Because what if you have a sequence of a sequence of linear that kind of thing?

File: spmd/tensor/parallel/utils.py (hunk 13 above)
Comment (fduwjj) [reply spmd/tensor/parallel/utils.py:48]: For now we can skip this case right? Or you feel this is important to handle?

File: spmd/tensor/parallel/style.py (hunk 1 above)
Comment (fduwjj) [reply spmd/tensor/parallel/style.py:21]: Because if you see ColwiseParallel, the prepare_output is None.

File: spmd/tensor/parallel/style.py (hunk 14 above)
Comment (fduwjj) [reply spmd/tensor/parallel/style.py:37]: Changed the comment here. But I would not use the world MLP since L means random layers of perceptrons.

File: spmd/tensor/parallel/utils.py (hunk 13 above)
Comment (fduwjj) [reply spmd/tensor/parallel/utils.py:48]: Addressed offline.

File: spmd/tensor/parallel/style.py (hunk 1 above)
Comment (fduwjj) [reply spmd/tensor/parallel/style.py]: Removed.

**Review by wanchaol — APPROVED** [review]
lgtm, thanks for addressing the comments!

**Comment by XilunWu** [comment]
landing of #622 causes conflict. sorry for that.
